File: apps/ios-dev/ClaudeContext/Sources/Services/Session/SessionManager.swift
import Foundation
import Combine

/// Manages AI conversation sessions with persistence, branching, restore, and cross-session context sharing.
@MainActor
class SessionManager: ObservableObject {

    @Published var state = SessionManagerState()

    private var sessionsDir: URL?
    private let maxRecentSessions = 20
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    init() {
        sessionsDir = nil
        encoder.outputFormatting = .prettyPrinted
    }

    func configure(projectPath: String) {
        let dir = URL(fileURLWithPath: projectPath).appendingPathComponent(".claude-context/sessions")
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        sessionsDir = dir
        loadSessions()
    }

    // MARK: - Session Lifecycle

    @discardableResult
    func createSession(title: String = "New Session", mode: AiMode = .agent, projectId: UUID? = nil, systemPrompt: String? = nil, modelId: String = "claude-sonnet-4-20250514") -> Session {
        let session = Session(
            title: title,
            projectId: projectId,
            mode: mode,
            metadata: SessionMetadata(modelId: modelId, systemPrompt: systemPrompt)
        )

        state.sessions.append(session)
        state.activeSessionId = session.id
        state.recentSessions = ([session] + state.recentSessions.filter { $0.id != session.id })
            .prefix(maxRecentSessions).map { $0 }

        persistSession(session)
        return session
    }

    func switchSession(_ sessionId: UUID) {
        guard let session = state.sessions.first(where: { $0.id == sessionId }) else { return }
        state.activeSessionId = sessionId
        state.recentSessions = ([session] + state.recentSessions.filter { $0.id != sessionId })
            .prefix(maxRecentSessions).map { $0 }
    }

    var activeSession: Session? {
        guard let id = state.activeSessionId else { return nil }
        return state.sessions.first { $0.id == id }
    }

    // MARK: - Message Management

    @discardableResult
    func addMessage(_ sessionId: UUID, message: ChatMessage) -> Session? {
        return updateSession(sessionId) { session in
            var s = session
            if let branchId = session.activeBranchId,
               let idx = session.branches.firstIndex(where: { $0.id == branchId }) {
                s.branches[idx].messages.append(message)
            } else {
                s.messages.append(message)
            }
            s.updatedAt = Date()
            return s
        }
    }

    func getMessages(_ sessionId: UUID) -> [ChatMessage] {
        guard let session = state.sessions.first(where: { $0.id == sessionId }) else { return [] }
        if let branchId = session.activeBranchId,
           let branch = session.branches.first(where: { $0.id == branchId }) {
            return Array(session.messages.prefix(branch.forkPointMessageIndex)) + branch.messages
        }
        return session.messages
    }

    // MARK: - Branching

    @discardableResult
    func createBranch(_ sessionId: UUID, name: String, forkAt: Int? = nil) -> SessionBranch? {
        guard let session = state.sessions.first(where: { $0.id == sessionId }) else { return nil }
        let forkPoint = forkAt ?? session.messages.count

        let branch = SessionBranch(name: name, forkPointMessageIndex: forkPoint, parentBranchId: session.activeBranchId)

        updateSession(sessionId) { s in
            var updated = s
            updated.branches.append(branch)
            updated.activeBranchId = branch.id
            updated.updatedAt = Date()
            return updated
        }

        return branch
    }

    func switchBranch(_ sessionId: UUID, branchId: UUID?) {
        updateSession(sessionId) { s in
            var updated = s
            updated.activeBranchId = branchId
            updated.updatedAt = Date()
            return updated
        }
    }

    @discardableResult
    func mergeBranch(_ sessionId: UUID, branchId: UUID) -> Session? {
        return updateSession(sessionId) { session in
            guard let branch = session.branches.first(where: { $0.id == branchId }) else { return session }
            var s = session
            let baseMessages = Array(session.messages.prefix(branch.forkPointMessageIndex))
            s.messages = baseMessages + branch.messages
            s.branches.removeAll { $0.id == branchId }
            if s.activeBranchId == branchId { s.activeBranchId = nil }
            s.updatedAt = Date()
            return s
        }
    }

    // MARK: - Checkpoints

    @discardableResult
    func createCheckpoint(_ sessionId: UUID, name: String, note: String = "") -> SessionCheckpoint? {
        guard let session = state.sessions.first(where: { $0.id == sessionId }) else { return nil }
        let checkpoint = SessionCheckpoint(name: name, messageIndex: session.messages.count, note: note)

        updateSession(sessionId) { s in
            var updated = s
            updated.checkpoints.append(checkpoint)
            updated.updatedAt = Date()
            return updated
        }

        return checkpoint
    }

    @discardableResult
    func restoreCheckpoint(_ sessionId: UUID, checkpointId: UUID) -> Session? {
        return updateSession(sessionId) { session in
            guard let checkpoint = session.checkpoints.first(where: { $0.id == checkpointId }) else { return session }
            var s = session
            s.messages = Array(session.messages.prefix(checkpoint.messageIndex))
            s.updatedAt = Date()
            return s
        }
    }

    // MARK: - Session Operations

    func updateTitle(_ sessionId: UUID, title: String) {
        updateSession(sessionId) { s in var u = s; u.title = title; u.updatedAt = Date(); return u }
    }

    func addTag(_ sessionId: UUID, tag: String) {
        updateSession(sessionId) { s in
            var u = s
            if !u.tags.contains(tag) { u.tags.append(tag) }
            u.updatedAt = Date()
            return u
        }
    }

    func archiveSession(_ sessionId: UUID) {
        updateSession(sessionId) { s in var u = s; u.status = .archived; u.updatedAt = Date(); return u }
    }

    func deleteSession(_ sessionId: UUID) {
        state.sessions.removeAll { $0.id == sessionId }
        state.recentSessions.removeAll { $0.id == sessionId }
        if state.activeSessionId == sessionId { state.activeSessionId = nil }
        if let dir = sessionsDir {
            try? FileManager.default.removeItem(at: dir.appendingPathComponent("\(sessionId).json"))
        }
    }

    @discardableResult
    func duplicateSession(_ sessionId: UUID) -> Session? {
        guard let original = state.sessions.first(where: { $0.id == sessionId }) else { return nil }
        let copy = createSession(title: "\(original.title) (copy)", mode: original.mode)
        return updateSession(copy.id) { s in
            var u = s
            u.messages = original.messages
            u.tags = original.tags
            u.parentSessionId = original.id
            return u
        }
    }

    // MARK: - Search

    func searchSessions(_ query: String) -> [Session] {
        let lower = query.lowercased()
        let results = state.sessions.filter { session in
            session.title.lowercased().contains(lower) ||
            session.tags.contains { $0.lowercased().contains(lower) } ||
            session.messages.contains { $0.content.lowercased().contains(lower) }
        }.sorted { $0.updatedAt > $1.updatedAt }

        state.searchResults = results
        return results
    }

    // MARK: - Export / Import

    func exportSession(_ sessionId: UUID) -> String? {
        guard let session = state.sessions.first(where: { $0.id == sessionId }),
              let data = try? encoder.encode(session) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    func importSession(_ json: String) -> Session? {
        guard let data = json.data(using: .utf8),
              var session = try? decoder.decode(Session.self, from: data) else { return nil }
        session.id = UUID()
        session.title = "\(session.title) (imported)"
        session.status = .active
        session.createdAt = Date()
        session.updatedAt = Date()
        state.sessions.append(session)
        persistSession(session)
        return session
    }

    // MARK: - Persistence

    private func loadSessions() {
        guard let dir = sessionsDir else { return }
        state.isLoading = true

        let files = (try? FileManager.default.contentsOfDirectory(at: dir, includingPropertiesForKeys: nil)
            .filter { $0.pathExtension == "json" }) ?? []

        let sessions: [Session] = files.compactMap { url in
            guard let data = try? Data(contentsOf: url) else { return nil }
            return try? decoder.decode(Session.self, from: data)
        }.sorted { $0.updatedAt > $1.updatedAt }

        state.sessions = sessions
        state.recentSessions = Array(sessions.prefix(maxRecentSessions))
        state.isLoading = false
    }

    private func persistSession(_ session: Session) {
        guard let dir = sessionsDir, let data = try? encoder.encode(session) else { return }
        try? data.write(to: dir.appendingPathComponent("\(session.id).json"))
    }

    @discardableResult
    private func updateSession(_ sessionId: UUID, transform: (Session) -> Session) -> Session? {
        guard let idx = state.sessions.firstIndex(where: { $0.id == sessionId }) else { return nil }
        let updated = transform(state.sessions[idx])
        state.sessions[idx] = updated
        persistSession(updated)
        return updated
    }
}

// MARK: - Data Models

struct Session: Identifiable, Codable {
    var id = UUID()
    var title: String = "New Session"
    var projectId: UUID?
    var mode: AiMode = .agent
    var messages: [ChatMessage] = []
    var branches: [SessionBranch] = []
    var activeBranchId: UUID?
    var metadata: SessionMetadata = SessionMetadata()
    var createdAt = Date()
    var updatedAt = Date()
    var status: SessionStatus = .active
    var tags: [String] = []
    var parentSessionId: UUID?
    var checkpoints: [SessionCheckpoint] = []
}

struct SessionBranch: Identifiable, Codable {
    let id = UUID()
    let name: String
    let forkPointMessageIndex: Int
    var parentBranchId: UUID?
    var messages: [ChatMessage] = []
    var createdAt = Date()
}

struct SessionMetadata: Codable {
    var modelId: String = "claude-sonnet-4-20250514"
    var systemPrompt: String?
    var temperature: Float = 0.7
    var maxTokens: Int = 4096
    var totalInputTokens: Int = 0
    var totalOutputTokens: Int = 0
    var estimatedCost: Double = 0.0
    var filesReferenced: [String] = []
}

struct SessionCheckpoint: Identifiable, Codable {
    let id = UUID()
    let name: String
    let messageIndex: Int
    var note: String = ""
    var createdAt = Date()
}

enum SessionStatus: String, Codable {
    case active, paused, completed, archived, failed
}

struct SessionManagerState {
    var sessions: [Session] = []
    var activeSessionId: UUID?
    var recentSessions: [Session] = []
    var searchResults: [Session] = []
    var isLoading = false
}

File: apps/android-dev/app/src/main/java/com/claudecontext/localdev/service/session/SessionManager.kt
package com.claudecontext.localdev.service.session

import com.claudecontext.localdev.data.models.*
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import java.io.File
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Manages AI conversation sessions with persistence, branching, restore,
 * and cross-session context sharing.
 */

// --- Data Models ---

data class Session(
    val id: String = UUID.randomUUID().toString(),
    val title: String = "New Session",
    val projectId: Long? = null,
    val mode: AiMode = AiMode.AGENT,
    val messages: List<ClaudeMessage> = emptyList(),
    val branches: List<SessionBranch> = emptyList(),
    val activeBranchId: String? = null,
    val metadata: SessionMetadata = SessionMetadata(),
    val createdAt: Long = System.currentTimeMillis(),
    val updatedAt: Long = System.currentTimeMillis(),
    val status: SessionStatus = SessionStatus.ACTIVE,
    val tags: List<String> = emptyList(),
    val parentSessionId: String? = null,
    val checkpoints: List<SessionCheckpoint> = emptyList()
)

data class SessionBranch(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val parentBranchId: String? = null,
    val forkPointMessageIndex: Int,
    val messages: List<ClaudeMessage> = emptyList(),
    val createdAt: Long = System.currentTimeMillis()
)

data class SessionMetadata(
    val modelId: String = "claude-sonnet-4-20250514",
    val systemPrompt: String? = null,
    val temperature: Float = 0.7f,
    val maxTokens: Int = 4096,
    val totalInputTokens: Long = 0,
    val totalOutputTokens: Long = 0,
    val estimatedCost: Double = 0.0,
    val filesReferenced: List<String> = emptyList(),
    val commandsExecuted: List<String> = emptyList()
)

data class SessionCheckpoint(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val messageIndex: Int,
    val note: String = "",
    val createdAt: Long = System.currentTimeMillis()
)

enum class SessionStatus {
    ACTIVE,
    PAUSED,
    COMPLETED,
    ARCHIVED,
    FAILED
}

data class SessionManagerState(
    val sessions: List<Session> = emptyList(),
    val activeSessionId: String? = null,
    val recentSessions: List<Session> = emptyList(),
    val searchResults: List<Session> = emptyList(),
    val isLoading: Boolean = false
)

@Singleton
class SessionManager @Inject constructor(
    private val gson: Gson
) {
    private val _state = MutableStateFlow(SessionManagerState())
    val state: StateFlow<SessionManagerState> = _state.asStateFlow()

    private var sessionsDir: File? = null
    private val maxRecentSessions = 20

    fun configure(projectPath: String) {
        sessionsDir = File(projectPath, ".claude-context/sessions").apply { mkdirs() }
        loadSessions()
    }

    // --- Session Lifecycle ---

    fun createSession(
        title: String = "New Session",
        mode: AiMode = AiMode.AGENT,
        projectId: Long? = null,
        systemPrompt: String? = null,
        modelId: String = "claude-sonnet-4-20250514"
    ): Session {
        val session = Session(
            title = title,
            projectId = projectId,
            mode = mode,
            metadata = SessionMetadata(
                modelId = modelId,
                systemPrompt = systemPrompt
            )
        )

        val sessions = _state.value.sessions + session
        _state.value = _state.value.copy(
            sessions = sessions,
            activeSessionId = session.id,
            recentSessions = (listOf(session) + _state.value.recentSessions).take(maxRecentSessions)
        )

        persistSession(session)
        return session
    }

    fun switchSession(sessionId: String) {
        val session = _state.value.sessions.find { it.id == sessionId } ?: return
        _state.value = _state.value.copy(
            activeSessionId = sessionId,
            recentSessions = (listOf(session) + _state.value.recentSessions.filter { it.id != sessionId })
                .take(maxRecentSessions)
        )
    }

    fun getActiveSession(): Session? {
        val id = _state.value.activeSessionId ?: return null
        return _state.value.sessions.find { it.id == id }
    }

    // --- Message Management ---

    fun addMessage(sessionId: String, message: ClaudeMessage): Session? {
        return updateSession(sessionId) { session ->
            val branchId = session.activeBranchId
            if (branchId != null) {
                val branches = session.branches.map { branch ->
                    if (branch.id == branchId) {
                        branch.copy(messages = branch.messages + message)
                    } else branch
                }
                session.copy(branches = branches, updatedAt = System.currentTimeMillis())
            } else {
                session.copy(
                    messages = session.messages + message,
                    updatedAt = System.currentTimeMillis()
                )
            }
        }
    }

    fun getMessages(sessionId: String): List<ClaudeMessage> {
        val session = _state.value.sessions.find { it.id == sessionId } ?: return emptyList()
        val branchId = session.activeBranchId
        if (branchId != null) {
            val branch = session.branches.find { it.id == branchId }
            if (branch != null) {
                return session.messages.take(branch.forkPointMessageIndex) + branch.messages
            }
        }
        return session.messages
    }

    fun editMessage(sessionId: String, messageIndex: Int, newContent: String): Session? {
        return updateSession(sessionId) { session ->
            val messages = session.messages.toMutableList()
            if (messageIndex in messages.indices) {
                messages[messageIndex] = messages[messageIndex].copy(content = newContent)
                // Remove all messages after the edit point and re-process
                val trimmed = messages.take(messageIndex + 1)
                session.copy(messages = trimmed, updatedAt = System.currentTimeMillis())
            } else session
        }
    }

    // --- Branching ---

    fun createBranch(sessionId: String, name: String, forkAtMessageIndex: Int? = null): SessionBranch? {
        val session = _state.value.sessions.find { it.id == sessionId } ?: return null
        val forkPoint = forkAtMessageIndex ?: session.messages.size

        val branch = SessionBranch(
            name = name,
            parentBranchId = session.activeBranchId,
            forkPointMessageIndex = forkPoint
        )

        updateSession(sessionId) { s ->
            s.copy(
                branches = s.branches + branch,
                activeBranchId = branch.id,
                updatedAt = System.currentTimeMillis()
            )
        }

        return branch
    }

    fun switchBranch(sessionId: String, branchId: String?) {
        updateSession(sessionId) { session ->
            session.copy(activeBranchId = branchId, updatedAt = System.currentTimeMillis())
        }
    }

    fun listBranches(sessionId: String): List<SessionBranch> {
        return _state.value.sessions.find { it.id == sessionId }?.branches ?: emptyList()
    }

    fun mergeBranch(sessionId: String, branchId: String): Session? {
        return updateSession(sessionId) { session ->
            val branch = session.branches.find { it.id == branchId } ?: return@updateSession session
            val baseMessages = session.messages.take(branch.forkPointMessageIndex)
            val mergedMessages = baseMessages + branch.messages

            session.copy(
                messages = mergedMessages,
                branches = session.branches.filter { it.id != branchId },
                activeBranchId = if (session.activeBranchId == branchId) null else session.activeBranchId,
                updatedAt = System.currentTimeMillis()
            )
        }
    }

    // --- Checkpoints ---

    fun createCheckpoint(sessionId: String, name: String, note: String = ""): SessionCheckpoint? {
        val session = _state.value.sessions.find { it.id == sessionId } ?: return null
        val checkpoint = SessionCheckpoint(
            name = name,
            messageIndex = session.messages.size,
            note = note
        )

        updateSession(sessionId) { s ->
            s.copy(
                checkpoints = s.checkpoints + checkpoint,
                updatedAt = System.currentTimeMillis()
            )
        }
        return checkpoint
    }

    fun restoreCheckpoint(sessionId: String, checkpointId: String): Session? {
        return updateSession(sessionId) { session ->
            val checkpoint = session.checkpoints.find { it.id == checkpointId }
                ?: return@updateSession session
            session.copy(
                messages = session.messages.take(checkpoint.messageIndex),
                updatedAt = System.currentTimeMillis()
            )
        }
    }

    // --- Session Operations ---

    fun updateSessionTitle(sessionId: String, title: String) {
        updateSession(sessionId) { it.copy(title = title, updatedAt = System.currentTimeMillis()) }
    }

    fun addTag(sessionId: String, tag: String) {
        updateSession(sessionId) { session ->
            if (tag !in session.tags) {
                session.copy(tags = session.tags + tag, updatedAt = System.currentTimeMillis())
            } else session
        }
    }

    fun removeTag(sessionId: String, tag: String) {
        updateSession(sessionId) { session ->
            session.copy(tags = session.tags - tag, updatedAt = System.currentTimeMillis())
        }
    }

    fun archiveSession(sessionId: String) {
        updateSession(sessionId) { it.copy(status = SessionStatus.ARCHIVED, updatedAt = System.currentTimeMillis()) }
    }

    fun deleteSession(sessionId: String) {
        val sessions = _state.value.sessions.filter { it.id != sessionId }
        val activeId = if (_state.value.activeSessionId == sessionId) null else _state.value.activeSessionId
        _state.value = _state.value.copy(
            sessions = sessions,
            activeSessionId = activeId,
            recentSessions = _state.value.recentSessions.filter { it.id != sessionId }
        )
        sessionsDir?.let { File(it, "$sessionId.json").delete() }
    }

    fun duplicateSession(sessionId: String): Session? {
        val original = _state.value.sessions.find { it.id == sessionId } ?: return null
        return createSession(
            title = "${original.title} (copy)",
            mode = original.mode,
            projectId = original.projectId,
            systemPrompt = original.metadata.systemPrompt,
            modelId = original.metadata.modelId
        ).let { newSession ->
            updateSession(newSession.id) { s ->
                s.copy(
                    messages = original.messages,
                    tags = original.tags,
                    parentSessionId = original.id
                )
            }
        }
    }

    // --- Search ---

    fun searchSessions(query: String): List<Session> {
        val lower = query.lowercase()
        val results = _state.value.sessions.filter { session ->
            session.title.lowercase().contains(lower) ||
            session.tags.any { it.lowercase().contains(lower) } ||
            session.messages.any { it.content.lowercase().contains(lower) }
        }.sortedByDescending { it.updatedAt }

        _state.value = _state.value.copy(searchResults = results)
        return results
    }

    fun getSessionsByTag(tag: String): List<Session> {
        return _state.value.sessions.filter { tag in it.tags }
    }

    fun getSessionsByProject(projectId: Long): List<Session> {
        return _state.value.sessions.filter { it.projectId == projectId }
            .sortedByDescending { it.updatedAt }
    }

    fun getSessionsByMode(mode: AiMode): List<Session> {
        return _state.value.sessions.filter { it.mode == mode }
    }

    // --- Export / Import ---

    suspend fun exportSession(sessionId: String): String? = withContext(Dispatchers.IO) {
        val session = _state.value.sessions.find { it.id == sessionId } ?: return@withContext null
        gson.toJson(session)
    }

    suspend fun importSession(json: String): Session? = withContext(Dispatchers.IO) {
        try {
            val session = gson.fromJson(json, Session::class.java)
            val imported = session.copy(
                id = UUID.randomUUID().toString(),
                title = "${session.title} (imported)",
                status = SessionStatus.ACTIVE,
                createdAt = System.currentTimeMillis(),
                updatedAt = System.currentTimeMillis()
            )
            val sessions = _state.value.sessions + imported
            _state.value = _state.value.copy(sessions = sessions)
            persistSession(imported)
            imported
        } catch (e: Exception) {
            null
        }
    }

    // --- Persistence ---

    private fun loadSessions() {
        val dir = sessionsDir ?: return
        _state.value = _state.value.copy(isLoading = true)

        val sessions = dir.listFiles { f -> f.extension == "json" }
            ?.mapNotNull { file ->
                try {
                    gson.fromJson(file.readText(), Session::class.java)
                } catch (e: Exception) {
                    null
                }
            }
            ?.sortedByDescending { it.updatedAt }
            ?: emptyList()

        _state.value = _state.value.copy(
            sessions = sessions,
            recentSessions = sessions.take(maxRecentSessions),
            isLoading = false
        )
    }

    private fun persistSession(session: Session) {
        val dir = sessionsDir ?: return
        try {
            File(dir, "${session.id}.json").writeText(gson.toJson(session))
        } catch (e: Exception) {
            // Log error
        }
    }

    private fun updateSession(sessionId: String, transform: (Session) -> Session): Session? {
        val sessions = _state.value.sessions.toMutableList()
        val index = sessions.indexOfFirst { it.id == sessionId }
        if (index < 0) return null

        val updated = transform(sessions[index])
        sessions[index] = updated

        _state.value = _state.value.copy(sessions = sessions)
        persistSession(updated)
        return updated
    }

    // --- Statistics ---

    fun getStats(): SessionStats {
        val sessions = _state.value.sessions
        val totalMessages = sessions.sumOf { it.messages.size }
        val totalCost = sessions.sumOf { it.metadata.estimatedCost }

        return SessionStats(
            totalSessions = sessions.size,
            activeSessions = sessions.count { it.status == SessionStatus.ACTIVE },
            totalMessages = totalMessages,
            totalBranches = sessions.sumOf { it.branches.size },
            totalCheckpoints = sessions.sumOf { it.checkpoints.size },
            estimatedTotalCost = totalCost,
            mostUsedMode = sessions.groupBy { it.mode }
                .maxByOrNull { it.value.size }?.key ?: AiMode.AGENT,
            sessionsByDay = sessions.groupBy {
                java.text.SimpleDateFormat("yyyy-MM-dd").format(java.util.Date(it.createdAt))
            }.mapValues { it.value.size }
        )
    }
}

data class SessionStats(
    val totalSessions: Int,
    val activeSessions: Int,
    val totalMessages: Int,
    val totalBranches: Int,
    val totalCheckpoints: Int,
    val estimatedTotalCost: Double,
    val mostUsedMode: AiMode,
    val sessionsByDay: Map<String, Int>
)
